### PR TITLE
GNN asset clustering

### LINKED ISSUE #59: Asset Clustering using Graph Neural Networks
- Request to create a pull request for Asset Clustering using Graph Neural Networks.

### PR DESCRIPTION
**Pull request by tzemicheal**
## Description

Closes #59 
## Checklist
- [X] I am familiar with the [Contributing Guidelines](https://github.com/nv-morpheus/Morpheus/blob/main/docs/source/developer_guide/contributing.md).
- [X] New or existing tests cover these changes.
- [X] The documentation is up to date with these changes.

### CODE REVIEW

File: gnn-asset-clustering/training-tuning-inference/data_loader.py [hunk 1]
```diff
@@ -0,0 +1,395 @@
...
+        # std_host_sys_port = sys_ports_df['SRC_PORT'].std() if sys_ports_df.shape[0] > 0 else np.nan
+        host_freq_sys_port = sys_ports_df['SRC_PORT'].mode()[0] if sys_ports_df.shape[0] > 0 else np.nan
+        # count_host_freq_sys_port = sys_ports_df['SRC_PORT'].value_counts().values.tolist()[0] if sys_ports_df.shape[0] > 0 else np.nan
+
+        user_ports_df = mac_df[mac_df['SRC_PORT'].between(1024, 49151, inclusive='both')] if \
+        mac_df[mac_df['SRC_PORT'].between(1024, 49151, inclusive='both')].shape[0] > 0 else pd.DataFrame()
+        count_host_no_of_user_ports = user_ports_df['SRC_PORT'].nunique() if user_ports_df.shape[0] > 0 else np.nan
+        # std_host_user_port = user_ports_df['SRC_PORT'].std() if user_ports_df.shape[0] > 0 else np.nan
+        host_freq_user_port = user_ports_df['SRC_PORT'].mode()[0] if user_ports_df.shape[0] > 0 else np.nan
+        # count_host_freq_user_port = user_ports_df['SRC_PORT'].value_counts().values.tolist()[0] if user_ports_df.shape[0] > 0 else np.nan
```
[the remark this answers is not in the document]
Comment (hsin-c) [reply gnn-asset-clustering/training-tuning-inference/data_loader.py:145]: Saw many of these down there too. If they're not used anymore, let's remove them all.

File: gnn-asset-clustering/training-tuning-inference/data_loader.py [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+        # pdus_device_mtype_armis = 1 if device_mtype=='PDUs' else np.nan
+        # singleBoardcomp_device_mtype_armis = 1 if device_mtype=='Single-Board Computers' else np.nan
+
+        device_os = device_armis_data.operatingSystem.iloc[0] if device_armis_data.shape[0] > 0 else ''
+        ubuntu_device_os_armis = 1 if device_os == 'Ubuntu' else np.nan
+        windows_device_os_armis = 1 if 'Windows' in device_os else np.nan
+        linux_device_os_armis = 1 if 'Linux' in device_os or 'Debian' in device_os or 'CentOS' in device_os else np.nan
+        vmwareESXi_device_os_armis = 1 if 'VMware' in device_os else np.nan
+
+        device_manufacturer = device_armis_data.manufacturer.iloc[0] if device_armis_data.shape[0] > 0 else ''
```
Comment (hsin-c) [inline gnn-asset-clustering/training-tuning-inference/data_loader.py:261]: I obfuscated all the device manufacturers' names so all of these values may need to be updated to reflect the new (obfuscated) values? The new values don't carry any meaning though (e.g. Computer Manufacturer 1). Do we want to keep the logic here? (Nothing in the obfuscated dataset will be matched though)

File: gnn-asset-clustering/training-tuning-inference/data_loader.py [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+        fujitsu_device_manufacturer_armis = 1 if 'Fujitsu' in device_manufacturer else np.nan
+        vmware_device_manufacturer_armis = 1 if 'VMware' in device_manufacturer else np.nan
+        arista_device_manufacturer_armis = 1 if 'Arista Networks' in device_manufacturer else np.nan
+        mellanox_device_manufacturer_armis = 1 if 'Mellanox Technologies' in device_manufacturer else np.nan
+        dell_device_manufacturer_armis = 1 if 'Dell' in device_manufacturer else np.nan
+        intel_device_manufacturer_armis = 1 if 'Intel' in device_manufacturer else np.nan
+
+        app_armis = armis_data_app.loc[armis_data_app.mac_address == raw_mac]
+        #print(app_armis.info())
+        app_device = app_armis.stack().groupby(level=0).apply(' '.join).iloc[0] if app_armis.shape[0] > 0 else ''
```
Comment (hsin-c) [inline gnn-asset-clustering/training-tuning-inference/data_loader.py:276]: Same issue here. All the app names are obfuscated (mapped to web-crawled apps from Azure Marketplace). So these value matchings don't work on the dataset anymore. How significant is the impact here?  What is the best way to fix it?

File: gnn-asset-clustering/training-tuning-inference/data_loader.py [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+        feat_dict = {
+            'raw_mac': raw_mac,
+            'mac_id': mac_id,
+            'count_of_host_sflows': count_of_host_sflows,
+            'count_host_no_of_sys_ports': count_host_no_of_sys_ports,  # 'std_host_sys_port' : std_host_sys_port,
+            'host_freq_sys_port': host_freq_sys_port,  # 'count_host_freq_sys_port' : count_host_freq_sys_port,
+            'count_host_no_of_user_ports': count_host_no_of_user_ports,  # 'std_host_user_port': std_host_user_port,
+            'host_freq_user_port': host_freq_user_port,  # 'count_host_freq_user_port': count_host_freq_user_port,
+            'count_host_no_of_dynamic_ports': count_host_no_of_dynamic_ports,
+            # 'count_other_no_of_sys_ports': count_other_no_of_sys_ports,
```
Comment (hsin-c) [inline gnn-asset-clustering/training-tuning-inference/data_loader.py:302]: Let's remove these if they're unused.

File: gnn-asset-clustering/training-tuning-inference/metrics.py [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+from sklearn.metrics import roc_auc_score
+from sklearn.metrics import average_precision_score
+from sklearn import metrics
+from munkres import Munkres
+import numpy as np
+from sklearn.metrics import silhouette_score, silhouette_samples
+import matplotlib.ticker as ticker
+
+
+class linkpred_metrics():
```
Comment (hsin-c) [inline gnn-asset-clustering/training-tuning-inference/metrics.py:25]: Can we use UpperCamelCase for class name here?

Comment (hsin-c) [reply gnn-asset-clustering/training-tuning-inference/metrics.py:25]: + some description of what the class is for would be super helpful

File: gnn-asset-clustering/training-tuning-inference/metrics.py [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+        return roc_score, ap_score, emb
+
+
+class ClusteringMetrics:
+
+    def __init__(self, true_label, predict_label):
+        self.true_label = true_label
+        self.pred_label = predict_label
+
+    def clusteringAcc(self):
```
Comment (hsin-c) [inline gnn-asset-clustering/training-tuning-inference/metrics.py:64]: Can we use snake_case for function names?

File: gnn-asset-clustering/training-tuning-inference/metrics.py [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+        acc = metrics.accuracy_score(self.true_label, new_predict)
+        f1_macro = metrics.f1_score(self.true_label, new_predict, average='macro')
+        precision_macro = metrics.precision_score(self.true_label, new_predict, average='macro')
+        recall_macro = metrics.recall_score(self.true_label, new_predict, average='macro')
+        f1_micro = metrics.f1_score(self.true_label, new_predict, average='micro')
+        precision_micro = metrics.precision_score(self.true_label, new_predict, average='micro')
+        recall_micro = metrics.recall_score(self.true_label, new_predict, average='micro')
+        return acc, f1_macro, precision_macro, recall_macro, f1_micro, precision_micro, recall_micro
+
+    def evaluationClusterModelFromLabel(self, print_msg=True):
```
Comment (hsin-c) [inline gnn-asset-clustering/training-tuning-inference/metrics.py:108]: let's use snake_case for function names

File: gnn-asset-clustering/training-tuning-inference/model.py [hunk 8]
```diff
@@ -0,0 +1,356 @@
...
+
+import torch
+import numpy as np
+from sklearn.cluster import KMeans
+from metrics import cal_clustering_metric, silhouette_plot
+import matplotlib.pyplot as plt
+
+
+class EGAE(torch.nn.Module):
+    """
```
Comment (hsin-c) [inline gnn-asset-clustering/training-tuning-inference/model.py:25]: Can we be a little more descriptive here for this class?

File: gnn-asset-clustering/training-tuning-inference/model.py [hunk 9]
```diff
@@ -0,0 +1,356 @@
...
+    X: n * d
+    """
+
+    def __init__(self,
```
Comment (hsin-c) [inline gnn-asset-clustering/training-tuning-inference/model.py:29]: A docstring for the inputs of the constructor will help readability by a lot

File: gnn-asset-clustering/training-tuning-inference/model.py [hunk 10]
```diff
@@ -0,0 +1,356 @@
...
+            recons_A = self(Laplacian)
+            loss = self.build_pretrain_loss(recons_A)
+            loss.backward()
+            optimizer.step()
+            # print(loss.item())
+        print(loss.item())
+
+
+class GAE(torch.nn.Module):
+    """
```
Comment (hsin-c) [inline gnn-asset-clustering/training-tuning-inference/model.py:225]: Let's be more descriptive and add a docstring for __init__

File: gnn-asset-clustering/training-tuning-inference/training-inference.ipynb
Comment (hsin-c) [inline gnn-asset-clustering/training-tuning-inference/training-inference.ipynb:1]: - Comment/class name mismatch:
  - "First, we define a Class 'radial' which...", but the class name is `Radar`. Is it intended?

Comment (tzemicheal) [reply gnn-asset-clustering/training-tuning-inference/training-inference.ipynb:1]: It is supposed to be "Radial class", fixed all over Radar reference.

File: gnn-asset-clustering/dataset/armis_enrichment_app_v3.csv
[the remark this answers is not in the document]
Comment (tzemicheal) [reply gnn-asset-clustering/dataset/armis_enrichment_app_v3.csv:1]: All are good.

File: gnn-asset-clustering/dataset/armis_enrichment_device_v3.csv
[the remark this answers is not in the document]
Comment (tzemicheal) [reply gnn-asset-clustering/dataset/armis_enrichment_device_v3.csv:1]: Yes, it is an obfuscated version.

File: gnn-asset-clustering/training-tuning-inference/data_loader.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (tzemicheal) [reply gnn-asset-clustering/training-tuning-inference/data_loader.py]: All codes refactored from the original version

File: gnn-asset-clustering/training-tuning-inference/metrics.py (hunk 5 above)
Comment (tzemicheal) [reply gnn-asset-clustering/training-tuning-inference/metrics.py]: Removed these two classes, as they are no longer used across any of the codes.

File: gnn-asset-clustering/training-tuning-inference/data_loader.py (hunk 2 above)
Comment (tzemicheal) [reply gnn-asset-clustering/training-tuning-inference/data_loader.py:227]: Let's keep this as it is, it doesn't have any impact if the device manufacturer's name is unavailable (the features assigned to empty). When the user provides data that has this value this still should work.

**Review by hsin-c — APPROVED** [review]

**Pull request merged**